### PR TITLE
Add `delete_record` to Registry API

### PR DESCRIPTION
**Pull request by ayushkamat**

### CODE REVIEW

File: latch/registry/table.py [hunk 1]
```diff
@@ -453,6 +473,71 @@ def _add_record_upserts_selection(self, updates: List[l.SelectionNode]) -> None:
...
+    def delete_record(self, name: str) -> None:
+        """Delete a record.
+
+        Args:
+            name: Target record name.
+        """
+        self._record_mutations.append(_TableRecordsDeleteData(name))
+
+    def _add_record_deletes_selection(
+        self, deletes: List[_TableRecordsDeleteData], updates: List[l.SelectionNode]
```
Comment (AidanAbd) [inline latch/registry/table.py:487]: Would rename updates to `mutations` to use the GQL language.

Comment (AidanAbd) [reply latch/registry/table.py:487]: Would do this everywhere.

File: latch/registry/table.py [hunk 2]
```diff
@@ -373,6 +389,8 @@ class TableUpdate:
 
     table: Table
 
+    # upsert record
```
Comment (AidanAbd) [inline latch/registry/table.py:392]: These boundary comments are unnecessary

File: latch/registry/table.py [hunk 3]
```diff
@@ -453,6 +473,71 @@ def _add_record_upserts_selection(self, updates: List[l.SelectionNode]) -> None:
...
+            if isinstance(next_mut, type(cur[0])):
+                cur.append(next_mut)
+                continue
+
+            output.append(cur)
+            cur = [next_mut]
+
+        output.append(cur)
+
+        return output
```
Comment (AidanAbd) [inline latch/registry/table.py:539]: Why is this necessary?

Comment (AidanAbd) [reply latch/registry/table.py:539]: Also I don't fully understand what this does -- in my mind this takes a list of [A,A,B,B,A,BBB] and turns it into [[A,A], [B,B], [A], [B,B,B]] -- not sure why this is useful.

File: latch/registry/table.py [hunk 4]
```diff
@@ -464,19 +549,27 @@ def commit(self) -> None:
...
         updates: List[l.SelectionNode] = []
 
-        self._add_record_upserts_selection(updates)
+        coalesced = self._coalesce_mutations()
+
+        for muts in coalesced:
+            if all(isinstance(mut, _TableRecordsUpsertData) for mut in muts):
+                self._add_record_upserts_selection(muts, updates)
+            if all(isinstance(mut, _TableRecordsDeleteData) for mut in muts):
+                self._add_record_deletes_selection(muts, updates)
```
Comment (AidanAbd) [inline latch/registry/table.py:558]: I see why you are using coalesce_mutations -- however I would just inline it. The code with the all(...) stuff is very weird even if it works in this case.

File: latch/registry/table.py [hunk 5]
```diff
@@ -464,19 +549,27 @@ def commit(self) -> None:
         """
         updates: List[l.SelectionNode] = []
```
Comment (AidanAbd) [inline latch/registry/table.py:550]: aren't these mutations not updates? I would go for strictly GQL language vs SQL language.

**Review by AidanAbd — CHANGES_REQUESTED** [review]

File: latch/registry/table.py (hunk 2 above)
Comment (ayushkamat) [reply latch/registry/table.py:392]: following the pattern set by max elsewhere in this api, idrc either way but it doesnt hurt

File: latch/registry/table.py (hunk 3 above)
Comment (ayushkamat) [reply latch/registry/table.py:539]: is useful because each string `[A, A, A, ..., A]` can be made into a single function call, so by coalescing in this way we reduce the size of the graphql query

File: latch/registry/table.py (hunk 5 above)
Comment (ayushkamat) [reply latch/registry/table.py:550]: sure

File: latch/registry/table.py (hunk 2 above)
Comment (maximsmol) [reply latch/registry/table.py:413]: well I had them around because the overload signatures take a full screen

File: latch/registry/table.py [hunk 6]
```diff
@@ -480,37 +494,110 @@ def _add_record_upserts_selection(
...
+        args = l.ArgumentNode()
+        args.name = _name_node("input")
+        args.value = _json_value(
+            {
+                "argExperimentId": self.table.id,
+                "argNames": names,
+            }
+        )
+
+        res.alias = _name_node(f"del{len(mutations)}")
```
Comment (maximsmol) [inline latch/registry/table.py:539]:
```suggestion
        res.alias = _name_node(f"upd{len(mutations)}")
```

File: latch/registry/table.py [hunk 7]
```diff
@@ -375,21 +380,27 @@ class _TableRecordsUpsertData:
...
+    _TableRecordsDeleteData,
+]
+
+
 @dataclass(frozen=True)
 class TableUpdate:
     """Ongoing :class:`Table` update transaction.
 
-    Groups requested updates to commit everything together in one network request.
+    Groups requested mutations to commit everything together in one network request.
```
Comment (maximsmol) [inline latch/registry/table.py:398]:
```suggestion
    Groups requested updates to commit everything together in one network request.
```

File: latch/registry/table.py [hunk 8]
```diff
@@ -480,37 +494,110 @@ def _add_record_upserts_selection(
...
+
+        return output
+
     def commit(self) -> None:
         """Commit this table update transaction.
 
         May be called multiple times.
 
-        All pending updates are committed with one network request.
+        All pending mutations are committed with one network request.
```
Comment (maximsmol) [inline latch/registry/table.py:573]:
```suggestion
        All pending updates are committed with one network request.
```

File: latch/registry/table.py [hunk 9]
```diff
@@ -480,37 +494,110 @@ def _add_record_upserts_selection(
...
+        res.arguments = tuple([args])
+
+        mutations.append(res)
+
+    # transaction
+
+    def _coalesce_mutations(
+        self,
+    ) -> List[Union[List[_TableRecordsUpsertData], List[_TableRecordsDeleteData]]]:
+        output = []
```
Comment (maximsmol) [inline latch/registry/table.py:549]: needs type

File: latch/registry/table.py [hunk 10]
```diff
@@ -480,37 +494,110 @@ def _add_record_upserts_selection(
...
-        updates: List[l.SelectionNode] = []
+        mutations: List[l.SelectionNode] = []
         vars: Dict[str, Tuple[l.TypeNode, JsonValue]] = {}
 
-        self._add_record_upserts_selection(updates, vars)
+        coalesced = self._coalesce_mutations()
 
-        if len(updates) == 0:
+        for muts in coalesced:
+            if all(isinstance(mut, _TableRecordsUpsertData) for mut in muts):
```
Comment (maximsmol) [inline latch/registry/table.py:583]: could just do one list traversal using like a generator

File: latch/registry/table.py [hunk 11]
```diff
@@ -527,8 +614,8 @@ def commit(self) -> None:
         self.clear()
 
     def clear(self):
-        """Remove pending updates.
+        """Remove pending mutations.
```
Comment (maximsmol) [inline latch/registry/table.py:617]:
```suggestion
        """Remove pending updates.
```

File: latch/registry/table.py [hunk 12]
```diff
@@ -527,8 +614,8 @@ def commit(self) -> None:
...
 
-        May be called to cancel any pending updates that have not been committed.
+        May be called to cancel any pending mutations that have not been committed.
```
Comment (maximsmol) [inline latch/registry/table.py:619]:
```suggestion
        May be called to cancel any pending updates that have not been committed.
```

File: latch/registry/utils.py [hunk 13]
```diff
@@ -76,6 +86,43 @@ def to_python_type(registry_type: RegistryType) -> Type[RegistryPythonValue]:
     )
 
 
+def to_registry_type(python_type: Type[RegistryPythonValue]) -> RegistryType:
```
Comment (maximsmol) [inline latch/registry/utils.py:89]: del

**Review by maximsmol — CHANGES_REQUESTED** [review]

File: latch/registry/table.py [hunk 14]
```diff
@@ -536,58 +536,49 @@ def _add_record_deletes_selection(
...
-            if all(isinstance(mut, _TableRecordsUpsertData) for mut in muts):
-                self._add_record_upserts_selection(muts, mutations, vars)
-            if all(isinstance(mut, _TableRecordsDeleteData) for mut in muts):
-                self._add_record_deletes_selection(muts, mutations)
-
-        if len(mutations) == 0:
+        if len(self._record_mutations) == 0:
             return
 
+        cur = []
```
[the remark this answers is not in the document]
Comment (maximsmol) [reply latch/registry/table.py:561]: move inside loop

File: latch/registry/table.py [hunk 15]
```diff
@@ -536,58 +536,49 @@ def _add_record_deletes_selection(
...
+        i = 0
+        while i < len(self._record_mutations):
+            j = i
+            while j < len(self._record_mutations):
```
Comment (maximsmol) [inline latch/registry/table.py:565]:
```suggestion
            for j in range(i, self._record_mutations):
```

File: latch/registry/table.py [hunk 16]
```diff
@@ -536,58 +536,49 @@ def _add_record_deletes_selection(
...
+                if not isinstance(
+                    self._record_mutations[j],
+                    type(self._record_mutations[i]),
+                ):
+                    break
+                cur.append(self._record_mutations[j])
+                j += 1
```
Comment (maximsmol) [inline latch/registry/table.py:572]:
```suggestion
                i = j
```

**Review by maximsmol — CHANGES_REQUESTED** [review]
little refactor

File: latch/registry/table.py (hunk 16 above)
Comment (ayushkamat) [reply latch/registry/table.py:572]: wont work bc of the isinstance check
